Make useless_ptr_null_checks smarter about some std functions

est31 opened the pull request:
This teaches the `useless_ptr_null_checks` lint that some std functions can't ever return null pointers, because they need to point to valid data, get references as input, etc.

This is achieved by introducing an `#[rustc_never_returns_null_ptr]` attribute and adding it to these std functions (gated behind bootstrap `cfg_attr`).

Later on, the attribute could maybe be used to tell LLVM that the returned pointer is never null. I don't expect much impact of that though, as the functions are pretty shallow and usually the input data is already never null.

Follow-up of PR #113657

Fixes #114442

est31 committed:
Add #[rustc_never_returns_null_ptr]

And look for it in the useless_ptr_null_checks lint

est31 committed:
Remove ptr_from_mut diagnostic item

It was added by #113657 for its purposes.
Now it is not used any more, remove it,
as we use the attr now.

Urgau commented on library/core/src/cell.rs:
```diff
@@ -2222,6 +2226,7 @@ impl<T: ?Sized> SyncUnsafeCell<T> {
     /// This call borrows the `SyncUnsafeCell` mutably (at compile-time) which
     /// guarantees that we possess the only reference.
     #[inline]
+    #[cfg_attr(not(bootstrap), rustc_never_returns_null_ptr)]
     pub const fn get_mut(&mut self) -> &mut T {
```
```suggestion
    pub const fn get_mut(&mut self) -> &mut T {
```
This function does not return a pointer.

Urgau commented on library/alloc/src/rc.rs:
```diff
@@ -2843,6 +2845,7 @@ impl<T: ?Sized, A: Allocator> Weak<T, A> {
     /// [`null`]: ptr::null
     #[must_use]
     #[stable(feature = "rc_as_ptr", since = "1.45.0")]
+    #[cfg_attr(not(bootstrap), rustc_never_returns_null_ptr)]
```
The documentation indicates that the pointer can be null:
> The pointer is valid only if there are some strong references. The pointer may be dangling,
> unaligned or even [`null`] otherwise.

but the code indicates otherwise as `Weak` internally uses `NonNull` :thinking:

Urgau commented on library/alloc/src/rc.rs:
```diff
@@ -2887,6 +2890,7 @@ impl<T: ?Sized, A: Allocator> Weak<T, A> {
     /// [`as_ptr`]: Weak::as_ptr
     #[must_use = "`self` will be dropped if the result is not used"]
     #[stable(feature = "weak_into_raw", since = "1.45.0")]
+    #[cfg_attr(not(bootstrap), rustc_never_returns_null_ptr)]
```
Same restrictions here, doc indicate that it may be null.

Urgau commented on library/alloc/src/sync.rs:
```diff
@@ -2625,6 +2627,7 @@ impl<T: ?Sized, A: Allocator> Weak<T, A> {
     /// [`null`]: core::ptr::null "ptr::null"
     #[must_use]
     #[stable(feature = "weak_into_raw", since = "1.45.0")]
+    #[cfg_attr(not(bootstrap), rustc_never_returns_null_ptr)]
```
Again here, doc indicates that the pointer may be null.

est31 committed:
Add #[rustc_never_returns_null_ptr] to std functions

Add the attribute to standard library functions that
are guaranteed to never return null pointers, as their
originating data wouldn't allow it.

est31 commented:
Good point about the function that returns a reference, I suppose I've missed that.

Also good point about the doc vs implementation discrepancy. I've removed the addition of the attribute from this PR and opened #114517 for discussion what to do about the discrepancy. It can then be resolved in a new PR.

est31 committed:
Improve diagnostics and add tests for function calls

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5773355530/job/15649180695) [(plain)](https://github.com/rust-lang/rust/commit/fed728d3007f2889e49896f68abcdfc62ddeab03/checks/15649180695/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:c85c95e3d7251135ab7dc9ce3241c5835cc595a9)
Download action repository 'actions/upload-artifact@v3' (SHA:0b7f8abb1508181956e8e162db84b466c27e18ce)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180116 sha256=351235b2326fb4db7a18e257e13ce7896c5f77339521e2c2612e71e154800a19
  Stored in directory: /tmp/pip-ephem-wheel-cache-6q7s1bjn/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
##[endgroup]
Built container sha256:b769ff9a561580d31b14c7b9eafec1faa9e9dd7c38b4ceaee5a5124621c4e627
Uploading finished image to https://ci-caches.rust-lang.org/docker/39289e18f1ac1b1a573d1a3c658528cab0cbfec4029e2def8ed615ecd4b1295e23213ab5b134f4a50a4bd45af0a8a4266b7571ba256dc217696d0f040efcd003

<botocore.awsrequest.AWSRequest object at 0x7ff1b93e5e10>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: /checkout/compiler/rustc_lint/messages.ftl: message `lint_ptr_null_checks_ref` appears before `lint_ptr_null_checks_fn_ret`, but is alphabetically later than it
run `./x.py test tidy --bless` to sort the file correctly
some tidy checks failed
##[group]Clock drift check
  local time: Sat Aug  5 23:44:18 UTC 2023
  network time: Sat, 05 Aug 2023 23:44:18 GMT
##[endgroup]
```

</details>

Urgau commented on compiler/rustc_lint/src/ptr_nulls.rs:
```diff
@@ -33,10 +33,25 @@ declare_lint_pass!(PtrNullChecks => [USELESS_PTR_NULL_CHECKS]);
 
 /// This function detects and returns the original expression from a series of consecutive casts,
 /// ie. `(my_fn as *const _ as *mut _).cast_mut()` would return the expression for `my_fn`.
```
```suggestion
/// This function checks if the expression is a from a series of consecutive casts,
/// ie. `(my_fn as *const _ as *mut _).cast_mut()` and if the original expression
/// is either a fn ptr, a reference or a function call to a
/// `rustc_never_returns_null_ptr` annotated function and if it is, returns the
/// appropriate dialog.
```

thomcc commented:
Sorry about the delay here. I think the std side is fine, but I can't review the compiler changes.

r? compiler

jackh726 commented:
@bors r+

bors commented:
:pushpin: Commit 4b1bc2701067d7b340d102d40227091c00447f48 has been approved by `jackh726`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jackh726 4b1bc2701067d7b340d102d40227091c00447f48 -->
<!-- homu: {"type":"Approved","sha":"4b1bc2701067d7b340d102d40227091c00447f48","approver":"jackh726","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 4b1bc2701067d7b340d102d40227091c00447f48 with merge 635c4a5e612b0ee8af6615635599702d3dce9906...
<!-- homu: {"type":"BuildStarted","head_sha":"4b1bc2701067d7b340d102d40227091c00447f48","merge_sha":"635c4a5e612b0ee8af6615635599702d3dce9906"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6205084615/job/16848838958)
Approved by: jackh726
Pushing 635c4a5e612b0ee8af6615635599702d3dce9906 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jackh726","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6205084615/job/16848838958"},"merge_sha":"635c4a5e612b0ee8af6615635599702d3dce9906"} -->

rust-timer commented:
Finished benchmarking commit (635c4a5e612b0ee8af6615635599702d3dce9906): [comparison URL](https://perf.rust-lang.org/compare.html?start=e81f85fe9eebf8a5ba21438fb188a59d54095697&end=635c4a5e612b0ee8af6615635599702d3dce9906&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=e81f85fe9eebf8a5ba21438fb188a59d54095697&end=635c4a5e612b0ee8af6615635599702d3dce9906&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e81f85fe9eebf8a5ba21438fb188a59d54095697&end=635c4a5e612b0ee8af6615635599702d3dce9906&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.3% | [3.3%, 3.3%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e81f85fe9eebf8a5ba21438fb188a59d54095697&end=635c4a5e612b0ee8af6615635599702d3dce9906&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=e81f85fe9eebf8a5ba21438fb188a59d54095697&end=635c4a5e612b0ee8af6615635599702d3dce9906&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 632.106s -> 632.885s (0.12%)
**Artifact size**: 318.13 MiB -> 318.13 MiB (0.00%)
